### ISSUE TITLE
Evolution des sites dans le temps

### ISSUE THREAD

**Issue by camillemonchicourt** [PnX-SI/gn_module_monitoring#132]
Déplacement du sujet de l'évolution des sites dans la temps dans un ticket dédié : 

## @Adrien-Pajot

Un point qui je pense va ici : la géométrie du site monitoré devrait pouvoir évoluer le temps, une demande dans le suivi de sites comme les habitats ou des colonies d'animaux ou de stations botaniques 

## @vprunet

De façon plus générale, nous avons identifié le besoin de faire évoluer les attributs d'un site dans le temps. Par exemple, un site de chiroptères peut être mis en protection, dégradé ou détruit, ou encore des aménagements favorables peuvent être réalisés sur le site.

Dans notre cas, il s'agit donc des attributs spécifiques au module.

J'ai imaginé faire cela en clonant le site à l'instant t-1 vers un nouveau site valide à partir de l'instant t, avec des attributs mis à jour et d'associer tous ces sites dans un même regroupement de sites.

Cette approche ne nécessite pas forcément d'évolution du modèle de données sql, chaque visite pointant sur le site correspondant au statut le jour de la visite. Eventuellement ajouter des attributs `first_use_date` et `last_use_date` à la table `gn_monitoring.t_site_complements` pour guider les observateurs (ou le logiciel) dans le choix du site à présenter à la saisie.

Mais je me rends compte que cette approche contrevient au principe de sites multi-modules. Créer un nouveau site n'a pas de sens alors que seul ses attributs relatifs à un module donné ont changé.

Une approche plus rigoureuse consisterait à conserver la notion de site partagé par tous les modules (hormis pour les changements liés au site même comme ses coordonnées) et créer un nouvel enregistrement dans la table `gn_monitoring.t_site_complements` en conservant les enregistrements précédents (plusieurs compléments en relation avec un site et un module donnés, chacun avec une période de validité distincte). Et idéalement, la visite pointerait non plus sur le site mais sur le site_complement en vigueur le jour de la visite pour le module (refonte non triviale des modèles éclatés entre geonature/data/core et  gn_monitoring/data/. Accessoirement id_module ne serait plus nécessaire dans t_base_visits et on serait sûr que le site pointé par la visite est ouvert pour le module, ce qui n'est pas le cas aujourd'hui avec la seule contrainte `fk_t_base_visits_id_base_site` ).

C'est un peu complexe, j'en conviens, mais cela rendrait le modèle de données plus puissant et plus robuste.

Sur le front end pour gérer la dimension temporelle des sites :

- faciliter la création du nouveau site (ou site complément)  comme une copie à modifier du site (ou complément) précédent.
- ouvrir à la saisie de visites les seuls sites (ou compléments) actifs à la date de visite.
- présenter un historique de l'évolution d'un site (ou des compléments du site) pour le module.

**Comment by DonovanMaillard** [comment]
Pourquoi ne pas prévoir une notion générique d' "evenement" ? Qui permettrait de stocker un ou des evenements rattachés à un objet (site, module, visite, obs... individu pour une cmr etc) dans une table transversale ? Ca permettrait d'avoir pour un site un certain nombre devenements, possiblement propres à chaque modules par exemple.

**Comment by vprunet** [comment]
@DonovanMaillard 
Oui, cela répondrait à mon besoin du moment que l'on peut facilement retrouver l'état d'un site à l'instant t (soit qu'il ait été enregistré tel quel, soit qu'on rejoue, à l'endroit ou à l'envers, une liste d'évènements relatifs au site), sachant par ailleurs que le modèle conceptuel de l'interface utilisateur est indépendant du modèle de données (un évènement peut  être synthétisé comme la différence entre l'état initial et modifié d'un formulaire).
